File: vaip_pass_level1_dd_cxx/src/partitioner.cpp
/*
 *     The Xilinx Vitis AI Vaip in this distribution are provided under the
 * following free and permissive binary-only license, but are not provided in
 * source code form.  While the following free and permissive license is similar
 * to the BSD open source license, it is NOT the BSD open source license nor
 * other OSI-approved open source license.
 *
 *      Copyright (C) 2023 – 2024 Advanced Micro Devices, Inc. All rights
 * reserved.
 *
 *      Redistribution and use in binary form only, without modification, is
 * permitted provided that the following conditions are met:
 *
 *      1. Redistributions must reproduce the above copyright notice, this list
 * of conditions and the following disclaimer in the documentation and/or other
 * materials provided with the distribution.
 *
 *      2. The name of Xilinx, Inc. may not be used to endorse or promote
 * products redistributed with this software without specific prior written
 * permission.
 *
 *      THIS SOFTWARE IS PROVIDED BY XILINX, INC. "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL XILINX, INC. BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *      PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE
 */
#include "dd.hpp"
#include <algorithm>
#include <iostream>

namespace dd {

void print_node_list(std::string str, node_list& nodes) {
  std::cout << str << "[";
  for (auto node : nodes)
    std::cout << node << ", ";
  std::cout << "]" << std::endl;
}
void print_property_map(std::string str, property_map& node_labels) {
  std::cout << str << "{";
  for (auto l : node_labels) {
    std::cout << l.first << ": " << l.second << ", ";
  }
  std::cout << "}" << std::endl;
}
void print_label_map(std::string str, label_map& labels) {
  std::cout << str << "{";
  for (auto l : labels) {
    std::cout << l.first << ": " << l.second << ", ";
  }
  std::cout << "}" << std::endl;
}
void print_graph(std::string str, Graph& g) {
  std::cout << str << "{";
  for (auto node : g) {
    std::cout << node.first << ": [";
    for (auto ind : node.second)
      std::cout << " " << ind;
    std::cout << "], ";
  }
  std::cout << "}" << std::endl;
}
node_list unique(node_list& in_nodes) {
  node_set temp_set(in_nodes.begin(), in_nodes.end());
  node_list out_nodes(temp_set.begin(), temp_set.end());
  return out_nodes;
}
node_list stable_unique(node_list& in_nodes) {
  std::map<node_ind_t, node_ind_t> temp_map;
  node_list out_nodes;
  for (auto node : in_nodes) {
    auto it = temp_map.insert({node, 0});
    if (it.second)
      out_nodes.emplace_back(node);
  }
  return out_nodes;
}
CompositeGraph::CompositeGraph(Graph adj_graph) {
  child_graph_ = adj_graph;
  // print_graph("Child_graph : ", child_graph_);
  child_graph_to_parent_graph(adj_graph);
  for (auto node : child_graph_) {
    labels_.insert(std::make_pair(node.first, node.first));
  }
  for (auto l : labels_) {
    clusters_.insert(std::make_pair(l.second, node_list()));
    clusters_[l.second].emplace_back(l.first);
  }
  get_input_nodes();
}

void CompositeGraph::child_graph_to_parent_graph(Graph adj_graph) {
  for (auto node : child_graph_) {
    parent_graph_.insert(std::make_pair(node.first, node_list()));
    for (auto child : node.second) {
      parent_graph_.insert(std::make_pair(child, node_list()));
      parent_graph_[child].emplace_back(node.first);
    }
  }
  // print_graph("Parent_graph : ", parent_graph_);
}
void CompositeGraph::get_input_nodes() {
  for (auto node : parent_graph_) {
    if (node.second.size() == 0)
      input_nodes_.emplace_back(node.first);
  }
}

node_ind_t CompositeGraph::get_subgraph_label_of_node(node_ind_t node) {
  return labels_[node];
}

node_ind_t CompositeGraph::get_subgraph_for_node(node_ind_t node) {
  auto label = get_subgraph_label_of_node(node);
  return label;
}

node_list CompositeGraph::get_next_nodes_of_node(node_ind_t node) {
  return child_graph_[node];
}

node_list CompositeGraph::get_input_subgraphs() {
  node_list subgs;
  for (auto node : input_nodes_)
    subgs.emplace_back(get_subgraph_label_of_node(node));
  auto unique_subgs = stable_unique(subgs);
  return unique_subgs;
}

node_list CompositeGraph::get_nodes_in_subgraph(node_ind_t subg) {
  auto nodes = clusters_[subg];
  return stable_unique(nodes);
}

node_list CompositeGraph::get_next_subgraphs(node_ind_t subg) {
  auto nodes = get_nodes_in_subgraph(subg);
  node_list next_nodes;
  for (auto node : nodes) {
    auto child_nodes = child_graph_[node];
    for (auto child_node : child_nodes) {
      if (std::find(nodes.begin(), nodes.end(), child_node) == nodes.end())
        next_nodes.emplace_back(child_node);
    }
  }
  node_list next_subgs;
  for (auto node : next_nodes) {
    next_subgs.emplace_back(get_subgraph_label_of_node(node));
  }
  next_subgs = stable_unique(next_subgs);
  return next_subgs;
}

bool CompositeGraph::is_cycle_detected(node_ind_t subg, node_set& subgs_visited,
                                       node_set& subgs_in_stack) {
  subgs_visited.insert(subg);
  subgs_in_stack.insert(subg);
  auto next_subgs = get_next_subgraphs(subg);
  for (auto next_subg : next_subgs) {
    if (subgs_in_stack.find(next_subg) != subgs_in_stack.end())
      return true;
    if (subgs_visited.find(next_subg) == subgs_visited.end())
      if (is_cycle_detected(next_subg, subgs_visited, subgs_in_stack))
        return true;
  }
  subgs_in_stack.erase(subg);
  return false;
}

bool CompositeGraph::is_dag() {
  node_set subgs_visited;
  node_set subgs_in_stack;

  auto input_subgs = get_input_subgraphs();
  for (auto subg : input_subgs) {
    if (is_cycle_detected(subg, subgs_visited, subgs_in_stack))
      return false;
  }
  return true;
}

void CompositeGraph::fuse(node_ind_t subgraph1, node_ind_t subgraph2) {
  if (subgraph1 < subgraph2) {
    auto subgraph2_nodes = clusters_[subgraph2];
    clusters_[subgraph1].insert(clusters_[subgraph1].end(),
                                subgraph2_nodes.begin(), subgraph2_nodes.end());
    for (auto node : subgraph2_nodes) {
      labels_[node] = subgraph1;
    }
  } else if (subgraph2 < subgraph1) {
    auto subgraph1_nodes = clusters_[subgraph1];
    clusters_[subgraph2].insert(clusters_[subgraph2].end(),
                                subgraph1_nodes.begin(), subgraph1_nodes.end());
    for (auto node : subgraph1_nodes) {
      labels_[node] = subgraph2;
    }
  }
}

void CompositeGraph::fuse_all(node_list subgraphs) {}

bool CompositeGraph::try_fuse(node_ind_t subgraph1, node_ind_t subgraph2) {
  label_map old_labels(labels_);
  auto subg1_nodes = clusters_[subgraph1];
  auto subg2_nodes = clusters_[subgraph2];

  fuse(subgraph1, subgraph2);
  if (is_dag()) {
    return true;
  } else {
    labels_ = old_labels;
    clusters_[subgraph1] = subg1_nodes;
    clusters_[subgraph2] = subg2_nodes;
    return false;
  }
}

node_list CompositeGraph::topsort() {
  std::vector<node_ind_t> s; //(input_nodes_.begin(), input_nodes_.end());
  node_list result;

  for (auto node : input_nodes_)
    s.push_back(node);
  node_set visited_nodes(input_nodes_.begin(), input_nodes_.end());
  while (!s.empty()) {
    auto node = s.back();
    auto child_nodes = get_next_nodes_of_node(node);
    for (auto child_node : child_nodes) {
      if (visited_nodes.find(child_node) == visited_nodes.end()) {
        s.push_back(child_node);
        visited_nodes.insert(child_node);
      }
    }
    if (s.back() == node) {
      result.emplace_back(node);
      s.pop_back();
    }
  }
  node_list res(result.rbegin(), result.rend());
  return res;
}

label_map partition_graph(Graph adj_graph, property_map property,
                          std::string optimization_flag,
                          const std::vector<size_t>& sorted_node_ids) {
  std::vector<std::string> optim_flags = {"L0", "L1", "L2"};
  assert(std::find(optim_flags.begin(), optim_flags.end(), optimization_flag) !=
         optim_flags.end());

  auto graph = CompositeGraph(adj_graph);

  if (optimization_flag == "L0")
    return graph.labels_;

  node_list nodes;
  bool sorted = !sorted_node_ids.empty();
  if (sorted) {
    for (auto& item : sorted_node_ids)
      nodes.push_back((int32_t)item);
  } else
    nodes = graph.topsort();

  for (auto i = 0; i < ((int32_t)nodes.size() - 1); i++) {
    auto node = nodes[i];
    auto next_node = nodes[i + 1];
    auto node_property = property[node];
    if (node_property != "CPU" && (node_property == property[next_node])) {
      auto subg_i = graph.get_subgraph_for_node(node);
      auto subg_in = graph.get_subgraph_for_node(next_node);
      graph.fuse(subg_i, subg_in);
    }
  }
  std::queue<node_ind_t> q;
  node_set visited_nodes;
  // L1 Partition
  for (auto node : graph.input_nodes_) {
    q.push(node);
    visited_nodes.insert(node);
  }

  while (!q.empty()) {
    auto node = q.front();
    q.pop();
    auto try_fuse = property[node] != "CPU" ? true : false;
    auto sg = graph.get_subgraph_for_node(node);
    auto next_nodes = graph.get_next_nodes_of_node(node);
    for (auto next_node : next_nodes) {
      if (try_fuse && (property[node] == property[next_node])) {
        auto next_sg = graph.get_subgraph_for_node(next_node);
        // auto status = graph.try_fuse(sg, next_sg);
        graph.try_fuse(sg, next_sg);
      }
      if (visited_nodes.find(next_node) == visited_nodes.end()) {
        q.push(next_node);
        visited_nodes.insert(next_node);
      }
    }
  }
  if (optimization_flag == "L1")
    return graph.labels_;

  if (optimization_flag == "L2")
    throw std::runtime_error(
        "L2 optimization is not implmented in Graph Partitioner");

  return graph.labels_;
}

Graph subgraph_labels_to_clusters(label_map subgraphs) {
  std::set<node_ind_t> cluster_inds;
  for (auto node : subgraphs)
    cluster_inds.insert(node.second);
  Graph clusters;
  for (auto c : cluster_inds)
    clusters.insert(std::make_pair(c, node_list()));
  for (auto node : subgraphs)
    clusters[node.second].emplace_back(node.first);
  return clusters;
}

} // namespace dd

File: vaip/python/voe/passes/partitioner.py
##
##  Copyright (C) 2023 – 2024 Advanced Micro Devices, Inc.
##  Licensed under the MIT License.
##
##  Licensed under the Apache License, Version 2.0 (the "License");
##  you may not use this file except in compliance with the License.
##  You may obtain a copy of the License at
##
##  http://www.apache.org/licenses/LICENSE-2.0
##
##  Unless required by applicable law or agreed to in writing, software
##  distributed under the License is distributed on an "AS IS" BASIS,
##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##  See the License for the specific language governing permissions and
##  limitations under the License.
##
import collections
import os
import time


def unique(vector):
    return list(set(vector))


def stable_unique(vector):
    return list(dict.fromkeys(vector).keys())


def child_graph_to_parent_graph(graph):
    parents = {}
    for node, children in graph.items():
        parents.setdefault(node, [])
        for child in children:
            parents.setdefault(child, []).append(node)
    return parents


class SubGraph:
    def __init__(self, label, nodes=None, next_subgraphs=None):
        self.label = label
        self.nodes = nodes if nodes is not None else []
        self.next_subgraphs = next_subgraphs if next_subgraphs is not None else []

    def __repr__(self):
        return f"SubGraph (label={self.label}, nodes={self.nodes}, next_subs={self.next_subgraphs})"


class CompositeGraph:
    """Graph with a notion of subgraphs"""

    def __init__(self, adj_list):
        self.child_graph = adj_list
        self.parent_graph = child_graph_to_parent_graph(adj_list)

        # labels : map(node_id -> subgraph label it belongs to)
        self.labels = {node: node for node in adj_list.keys()}

        # clusters : map(subgraph label -> [nodes in subgraph])
        self.clusters = {label: [] for label in self.labels.values()}
        for node, label in self.labels.items():
            self.clusters[label].append(node)

    def get_input_nodes(self):
        return tuple(
            node
            for node, parent_nodes in self.parent_graph.items()
            if len(parent_nodes) == 0
        )

    def get_subgraph_label_of_node(self, node):
        return self.labels[node]

    def get_subgraph_for_node(self, node: int) -> int:
        label = self.get_subgraph_label_of_node(node)
        return label

    def get_next_nodes_of_node(self, node):
        return self.child_graph[node]

    def get_input_subgraphs(self) -> list[int]:
        """Return the list subgraphs which are inputs in the graph"""
        nodes = self.get_input_nodes()
        subgs = [self.get_subgraph_label_of_node(node) for node in nodes]
        unique_subgs = stable_unique(subgs)
        return unique_subgs

    def get_nodes_in_subgraph(self, subg: int) -> list[int]:
        """Given a subgraph, return the list of nodes in the subgraph"""
        nodes = self.clusters[subg]
        nodes = stable_unique(nodes)
        return nodes

    def get_next_subgraphs(self, subg: int) -> list[int]:
        """Given a subgraph, return its child subgraphs"""
        nodes = self.get_nodes_in_subgraph(subg)
        next_nodes = []
        for node in nodes:
            child_nodes = self.child_graph[node]
            for child_node in child_nodes:
                if child_node not in nodes:
                    next_nodes.append(child_node)
        next_subgs = [self.get_subgraph_label_of_node(node) for node in next_nodes]
        next_subgs = stable_unique(next_subgs)
        return next_subgs

    def is_cycle_detected(
        self, subg: int, subgs_visited: set[int], subgs_in_stack: set[int]
    ) -> bool:
        subgs_visited.add(subg)
        subgs_in_stack.add(subg)
        next_subgs = self.get_next_subgraphs(subg)
        for next_subg in next_subgs:
            if next_subg in subgs_in_stack:
                return True

            if next_subg not in subgs_visited:
                if self.is_cycle_detected(next_subg, subgs_visited, subgs_in_stack):
                    return True

        subgs_in_stack.remove(subg)
        return False

    def is_dag(self) -> bool:
        subgs_visited = set()
        subgs_in_stack = set()

        input_subgs = self.get_input_subgraphs()
        for subg in input_subgs:
            if self.is_cycle_detected(subg, subgs_visited, subgs_in_stack):
                return False
        return True

    def fuse(self, subgraph1: int, subgraph2: int):
        if subgraph1 < subgraph2:
            subgraph2_nodes = self.clusters[subgraph2]
            self.clusters[subgraph1].extend(subgraph2_nodes)
            for node in subgraph2_nodes:
                self.labels[node] = subgraph1
        elif subgraph2 < subgraph1:
            subgraph1_nodes = self.clusters[subgraph1]
            self.clusters[subgraph2].extend(subgraph1_nodes)
            for node in subgraph1_nodes:
                self.labels[node] = subgraph2
        else:
            pass

    def fuse_all(self, subgraphs: list[int]):
        smallest_label = min(subgraphs)
        other_subg_nodes = []
        for subgraph in subgraphs:
            if subgraph == smallest_label:
                continue
            other_subg_nodes.extend(self.clusters[subgraph])

        for node in other_subg_nodes:
            self.labels[node] = smallest_label
        self.clusters[smallest_label].extend(other_subg_nodes)

    def try_fuse(self, subgraph1: int, subgraph2: int) -> bool:
        # Back up existing setup
        old_labels = self.labels.copy()
        subg1_nodes = list(self.clusters[subgraph1])
        subg2_nodes = list(self.clusters[subgraph2])

        self.fuse(subgraph1, subgraph2)
        if self.is_dag():
            return True
        else:
            self.labels = old_labels.copy()
            self.clusters[subgraph1] = subg1_nodes
            self.clusters[subgraph2] = subg2_nodes
            return False

    def topsort(self):
        input_nodes = self.get_input_nodes()
        stack = []
        stack.extend(input_nodes)
        visited_nodes = set(input_nodes)
        result = []
        while stack:
            node = stack[-1]
            child_nodes = self.get_next_nodes_of_node(node)
            for child_node in child_nodes:
                if child_node not in visited_nodes:
                    stack.append(child_node)
                    visited_nodes.add(child_node)

            if stack[-1] == node:
                result.append(node)
                stack.pop()
        return result[::-1]


def test(graph):
    CGraph = CompositeGraph(graph)
    assert child_graph_to_parent_graph(graph)[0] == []
    assert CGraph.get_input_nodes() == (0,)
    assert CGraph.get_subgraph_label_of_node(1) == 1
    assert CGraph.get_next_nodes_of_node(0) == [1, 2]
    assert CGraph.get_input_subgraphs() == [0]
    assert CGraph.get_next_subgraphs(0) == [1, 2]
    assert CGraph.get_nodes_in_subgraph(1) == [1]
    assert CGraph.is_dag() == True

    CGraph.fuse(0, 1)
    assert 1 not in CGraph.labels.values()
    assert len(CGraph.get_next_subgraphs(0)) == 3
    assert CGraph.is_dag() == True
    print("After fusing 0 & 1 :", CGraph.labels)

    CGraph.fuse(0, 3)
    assert CGraph.is_dag() == True
    print("After fusing 0 & 3 :", CGraph.labels)

    assert CGraph.try_fuse(0, 7) == False
    print("After try_fusing 0 & 7 :", CGraph.labels)

    CGraph.fuse(0, 7)
    assert CGraph.is_dag() == False
    print("After fusing 0 & 7 :", CGraph.labels)

    print("TESTS PASSED")


def partition_graph(adj_graph, property, optimization_flag="L1", sorted=True):
    assert optimization_flag in {"L0", "L1", "L2"}

    queue = collections.deque()
    visited_nodes = set()

    graph = CompositeGraph(adj_graph)

    if optimization_flag == "L0":
        return graph.labels

    # Initial merging of subgraphs on topsorted list of nodes.
    nodes = graph.topsort() if not sorted else list(graph.labels.keys())
    for i, node in enumerate(nodes[:-1]):
        next_node = nodes[i + 1]
        node_property = property[node]
        if node_property != "CPU" and node_property == property[next_node]:
            subg_i = graph.get_subgraph_for_node(node)
            subg_in = graph.get_subgraph_for_node(next_node)
            graph.fuse(subg_i, subg_in)

    # L1 Partition
    for node in graph.get_input_nodes():
        queue.append(node)
        visited_nodes.add(node)

    while queue:
        node = queue.popleft()
        try_fuse = True if property[node] != "CPU" else False
        sg = graph.get_subgraph_for_node(node)
        next_nodes = graph.get_next_nodes_of_node(node)
        for next_node in next_nodes:
            if try_fuse and property[node] == property[next_node]:
                next_sg = graph.get_subgraph_for_node(next_node)
                status = graph.try_fuse(sg, next_sg)

            if next_node not in visited_nodes:
                queue.append(next_node)
                visited_nodes.add(next_node)

    if optimization_flag == "L1":
        return graph.labels

    if optimization_flag == "L2":
        raise RuntimeError("L2 optimization is not implmented in Graph Partitioner")


def subgraph_labels_to_clusters(subgraphs):
    clusters = dict.fromkeys(subgraphs.values())
    for key in clusters.keys():
        clusters[key] = []
    for node, label in subgraphs.items():
        clusters[label].append(node)
    return clusters


if __name__ == "__main__":
    from test_cases import tests

    test(tests[0]["adj_graph"])

    for test in tests:
        subgraphs = partition_graph(
            test["adj_graph"], test["property"], optimization_flag="L1"
        )
        clusters = subgraph_labels_to_clusters(subgraphs)
        print("Expected Result:", test["result"])
        print("Result Obtained:", clusters)
        print("-" * 80)
